### PR TITLE
Crossing INITs

### PR DESCRIPTION
**Pull request by AdityaSripal**
First pass at requested changes, they look good! cc: @damiannolan @colin-axner

### CODE REVIEW

File: spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md [hunk 1]
```diff
@@ -510,22 +505,14 @@ function chanUpgradeTry(
...
+    // both sides will have the same upgradeSequence
+    abortTransactionUnless(counterpartyUpgradeSequence > currentChannel.upgradeSequence)
 
-        initUpgradeChannelHandshake(portIdentifier, channelIdentifier, upgradeFields, counterpartyUpgrade.timeout)
-    } else if currentChannel.state == INITUPGRADE {
-        existingUpgrade = publicStore.get(channelUpgradePath)
-        abortTransactionUnless(existingUpgrade.fields == upgradeFields)
-    }
+    upgrade = publicStore.get(channelUpgradePath)
+    abortTransactionUnless(upgrade.fields == counterpartyUpgrade.fields)
```
Comment (damiannolan) [inline spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md:515]: Is this check necessary? Seems like a mistake - counterpartyUpgrade fields may have a different connection ID to upgrade fields connection ID and that should be valid, right? i.e. both connection ends do not need to share the same ID so we cannot make this assertion

File: spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md [hunk 2]
```diff
@@ -418,11 +418,14 @@ The upgrade handshake defines four datagrams: *ChanUpgradeInit*, *ChanUpgradeTry
...
-| Initiator | Datagram             | Chain acted upon | Prior state (A, B)            | Posterior state (A, B)    |
-| --------- | -------------------- | ---------------- | ----------------------------- | ------------------------- |
-| Actor     | `ChanUpgradeInit`    | A                | (OPEN, OPEN)                  | (INITUPGRADE, OPEN)       |
-| Actor     | `ChanUpgradeTry`     | B                | (INITUPGRADE, OPEN)           | (INITUPGRADE, TRYUPGRADE) |
-| Relayer   | `ChanUpgradeAck`     | A                | (INITUPGRADE, TRYUPGRADE)     | (ACKUPGRADE, TRYUPGRADE)  |
+| Initiator | Datagram             | Chain acted upon | Prior state (A, B)            | Posterior state (A, B)     |
+| --------- | -------------------- | ---------------- | ----------------------------- | -------------------------- |
+| Actor     | `ChanUpgradeInit`    | A                | (OPEN, OPEN)                  | (INITUPGRADE, OPEN)        |
+| Actor     | `ChanUpgradeInit`    | B                | (INITUPGRADE, OPEN)           | (INITUPGRADE, INITUPGRADE) |
+| Actor     | `ChanUpgradeTry`     | B                | (INITUPGRADE, OPEN)           | (INITUPGRADE, TRYUPGRADE)  |
```
Comment (damiannolan) [inline spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md:425]: should this be?

```suggestion
| Actor     | `ChanUpgradeTry`     | B                | (INITUPGRADE, INITUPGRADE)           | (INITUPGRADE, TRYUPGRADE)  |
```

File: spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md (hunk 1 above)
Comment (AdityaSripal) [reply spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md]: Yes you're right. The necessary checks are made in `startFlushHandshake`

File: spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md [hunk 3]
```diff
@@ -418,19 +418,25 @@ The upgrade handshake defines four datagrams: *ChanUpgradeInit*, *ChanUpgradeTry
...
 
 Once both states are in `ACKUPGRADE` and `TRYUPGRADE` respectively, both sides must move to `FLUSHINGCOMPLETE` respectively by clearing their in-flight packets. Once both sides have complete flushing, a relayer may submit a `ChanUpgradeOpen` message to both ends proving that the counterparty has also completed flushing in order to move the channelEnd to `OPEN`.
 
-`ChanUpgradeOpen` is only necessary to call on chain A if the chain was not moved to `OPEN` on `ChanUpgradeAck` which may happen if all packets on both ends are already flushed.
+`ChanUpgradeOpen` can be called on both ends in either order once both sides have completed flushing. It may not be necessary on the ACK side if the ACK message automatically moves to `OPEN`.
 
 At the end of a successful upgrade handshake between two chains implementing the sub-protocol, the following properties hold:
 
-- Each chain is running their new upgraded channel end and is processing upgraded logic and state according to the upgraded parameters.
+- Each chain is running their new upgraded channel end is processing upgraded logic and state according to the upgraded parameters.
```
Comment (colin-axner) [inline spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md:439]: wording is off now

File: spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md [hunk 4]
```diff
@@ -510,22 +508,13 @@ function chanUpgradeTry(
...
-        // initUpgradeChannelHandshake will increment the sequence so after that call
-        // both sides will have the same upgradeSequence
-        if counterpartyUpgradeSequence > currentChannel.upgradeSequence {
-            currentChannel.upgradeSequence = counterpartyUpgradeSequence - 1
-        }
+    // if the counterparty sequence is greater than the current sequence, we fast forward to the counterparty sequence
+    // so that both channel ends are using the same sequence for the current upgrade
+    // initUpgradeChannelHandshake will increment the sequence so after that call
+    // both sides will have the same upgradeSequence
+    abortTransactionUnless(counterpartyUpgradeSequence > currentChannel.upgradeSequence)
```
Comment (colin-axner) [inline spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md:515]: This seems incorrect? code doc does not match the check and the check is using an unverified counterparty upgrade sequence, in addition it doesn't allow for the sequence to be equal?

File: spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md [hunk 5]
```diff
@@ -750,7 +738,7 @@ function cancelChannelUpgrade(
 ) {
     // current channel is in INITUPGRADE or TRYUPGRADE
     currentChannel = provableStore.get(channelPath(portIdentifier, channelIdentifier))
-    abortTransactionUnless(currentChannel.state == INITUPGRADE || currentChannel.state == TRYUPGRADE)
+    abortTransactionUnless(currentChannel.state != OPEN)
```
Comment (colin-axner) [inline spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md:741]: interesting, why did this change?

**Review by colin-axner — COMMENTED** [review]
Nice! From the ibc-go perspective, I'm definitely in favour. I know there was mentioning of this potentially increasing complexity for the relayer, if that's okay, ibc-go is fine to leave as is, but it'd be worth noting why it would increase complexity on the relayer? Maybe we can reason about this easier once we have nailed down concrete upgrade flows?

**Pull request closed without merging**